syntax = "proto3";

service Mafia {
    rpc ConnectToServer(ConnectToServerMessage) returns (EmptyServerResponse) {}
    rpc DisconnectFromServer(DisconnectFromServerMessage) returns (EmptyServerResponse) {}
    rpc GetConnectedUsers(GetConnectedUsersMessage) returns (GetConnectedUsersResponse) {}
    rpc GetNewMessage(GetMessageRequest) returns (GetMessageResponse) {}
    rpc SendUserCommand(SendUserCommandRequest) returns (EmptyServerResponse) {}
}

message ConnectToServerMessage {
    string session_name = 1;
    string user_name = 2;
}

message GetConnectedUsersMessage {
    string session_name = 1;
}

message DisconnectFromServerMessage {
    string session_name = 1;
    string user_name = 2;
}

message GetConnectedUsersResponse {
    repeated string user_names = 1;
}

message GetMessageRequest {
    string session_name = 1;
    string user_name = 2;
}

message GetMessageResponse {
    oneof message_type {
        NoneMessage none_message = 1;
        SuccessConnectionMessage success_connection = 2;
        NewConnectionMessage new_connection = 3;
        RemovedConnectionMessage removed_connection = 4;
        ChatMessage chat_message = 5;
    }

    message NoneMessage {}

    message SuccessConnectionMessage {
        repeated string current_users = 1;
    }

    message NewConnectionMessage {
        string new_user_name = 1;
        repeated string current_users = 2;
    }

    message RemovedConnectionMessage {
        string removed_user_name = 1;
        repeated string current_users = 2;
    }

    message ChatMessage {
        string user_name = 1;
        string text = 2;
    }
}

message EmptyServerResponse {}

message SendUserCommandRequest {
    oneof message_type {
        ChatMessage chat_message = 1;
    }

    message ChatMessage {
        string session_name = 1;
        string user_name = 2;
        string text = 3;
    }
}
